[NCCL PG] Enable storing nccl traces into storage and make it configurable

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113678
* __->__ #113503

This PR is to enable the store of NCCL flight recorder to storage and make it configurable by letting users register their own way of storing the debug info. We will then provide users a script to offline parse and process the dumped blobs.

One thing, this PR is not trying to resolve is to decide where to dump the debug info. I will send a follow-up PR to address that.


cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @wz337 @tianyu-l @wconstab @yf225 @kiukchung @d4l3k @lucasllc

fduwjj committed:
[NCCL PG] Enable storing nccl traces into storage and make it configurable

[ghstack-poisoned]

fduwjj committed:
Update on "[NCCL PG] Enable storing nccl traces into storage and make it configurable"




[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1276,13 +1277,46 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
+    auto ncclTrace = dump_nccl_trace();
+    if (debugInfoCallbackStorer_ != nullptr) {
+      debugInfoCallbackStorer_(rank_, ncclTrace);
+    } else {
```
nit: we could probably avoid having this else branch.

we could arrange for 'storeTraceInfoToLocalDisk' to be 'registered' by default (perhaps only in oss or in some settings).

fduwjj replied:
OK, originally I want the registration only happening once. But if we register it in watchdog thread, it will be called in the constructor. So maybe we can still do the null check here and register the default one if it is null.

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1276,13 +1277,46 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
+  // This can help in retrieving the strings back if needed.
+  size_t size = ncclTrace.size();
+  file.write(reinterpret_cast<const char*>(&size), sizeof(size));
+  file.write(ncclTrace.data(), size);
```
we might want to use zip writer here. i suppose that could be an optimization later, but how large are these files anyway?

fduwjj replied:
Well I guess it depends on how big we want to set for `TORCH_NCCL_TRACE_BUFFER_SIZE`? In the scenario of 16/32K training, for a typical LLM transformer, each layer will use about 2-4 collectives for FSDP, 4 for TP/SP and 0-2 for PP. So 10 collectives at most for one layer, and if we think looking back 3-5 layers should be suffice, then we probably only need 30-50 entries. If one entry take about 1KB(just random guess, maybe @zdevito can correct me if I am wrong). Then it would be 30-50 KB?

zdevito commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1276,13 +1277,46 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
-    LOG(ERROR) << "nccl_trace: " << dump_nccl_trace();
+    // We dump nccl trace into local disk by default and users can register
+    // their own storer via `registerDebugInfoCallbackStorer`.
+    auto ncclTrace = dump_nccl_trace();
```
This will not return on a GIL deadlock (it is going to gather python stack frames). Perhaps we should start a std::thread here just so we can timeout if we haven't finished writing it after like 20--30 seconds.

zdevito replied:
I guess since we are already going to wait later for stragglers we don't even have to wait for the thread to complete here, we could just report before we quit if the thread failed to finish writing the flight recorder.

fduwjj replied:
Yah, that make sense. I didn't know that it would hang there because of GIL deadlock. Is it possible that we can have a timeout for getting python stack frames so that we at least dump something or report there is a GIL deadlock?

I will use a separate thread here and check whether it is joinable or not in the later wait. And if it is not joinable, then we are going to sleep and mark the thread detachable. (The sleep will be 2 minutes, so it should be long enough)

zdevito replied:
How the code currently works is good to land. I think we can make the changes to the symbolizer to succeed even if it cannot acquire the GIL. In particular, I think we can it specify a timeout after which we either return without python frame info, or more sneakily, just assume the reason we didn't get the GIL for many seconds is because its deadlocked and just try gathering the python info without the lock, similar to what pyspy does anyway.

fduwjj replied:
@zdevito totally agree, I only mentioned this to @wconstab that we did need to return something even if there is GIL deadlock. I will follow up this in a separate PR.

wconstab replied:
> I will use a separate thread here and check whether it is joinable or not in the later wait

IIUC .joinable will return true any time after the thread actually gets created.  It does not mean "finished and can be joined now" but rather "is referring to a real OS thread", as opposed to, there is no real thread attached to this Thread obj"

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1276,13 +1277,46 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
 }
 
-void dump_debugging_info() {
+void storeTraceInfoToLocalDisk(int rank, const std::string& ncclTrace) {
```
i'm wondering if it is cleaner to define a 'class' for the disk dumper, so it can have a constructor that initializes things like its filename, and avoid processing envs in the actual write to disk method.

and then we'd also change the interface for registering from function to object following that interface.

i dont have a strong feeling here, maybe ok to just leave as function.

fduwjj replied:
I see. Yeah, making it a class would be better and later on users can just override the class and do their customized logic there.

XilunWu replied:
Let's also put this in `TestUtils.hpp`, WDYT?

fduwjj replied:
@XilunWu this is the default fallback function we will need to keep in this file. This is not for testing purpose at all.

wconstab commented on test/cpp/c10d/ProcessGroupNCCLErrorsTest.cpp:
```diff
@@ -346,23 +347,64 @@ TEST_F(ProcessGroupNCCLErrorsTest, testNCCLErrorsNonBlocking) {
+  // so that we can use it for validation.
+  pg.registerDebugInfoCallbackStorer(
+      [&](int rank, const std::string& ncclTrace) {
+        auto filename = c10::str("/tmp/test_nccl_trace_rank_", rank);
```
do we have a pattern in some other c++ test you can copy from? ideally we'd clean up the tmpfile after the test runs.  its easier in python.. but we can still do it

H-Huang replied:
can we generate the file through this? https://en.cppreference.com/w/cpp/io/c/tmpfile

wconstab commented on test/cpp/c10d/ProcessGroupNCCLErrorsTest.cpp:
```diff
@@ -346,23 +347,64 @@ TEST_F(ProcessGroupNCCLErrorsTest, testNCCLErrorsNonBlocking) {
 }
 
+// Function to read what we wrote to the local disk for validation.
+std::string readTraceInfoFromLocalDisk(int rank) {
```
nit: just make 'filename' an arg here instead of doing the rank formatting inside?

XilunWu replied:
yeah I agree, and we can put this util function in `TestUtils.hpp` as well as `#include <fstream>`.

fduwjj replied:
I can put the name as an input arg. For @XilunWu's point, I do feel this is kind of specific to this test. And we an always revisit it when we see more use cases for this function.

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1276,13 +1277,46 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
+  // Write the size of the string followed by the string itself.
+  // This can help in retrieving the strings back if needed.
+  size_t size = ncclTrace.size();
+  file.write(reinterpret_cast<const char*>(&size), sizeof(size));
```
ok.

H-Huang commented on test/cpp/c10d/ProcessGroupNCCLErrorsTest.cpp:
```diff
@@ -346,23 +347,64 @@ TEST_F(ProcessGroupNCCLErrorsTest, testNCCLErrorsNonBlocking) {
 
+// Function to read what we wrote to the local disk for validation.
+std::string readTraceInfoFromLocalDisk(int rank) {
+  auto filename = c10::str("/tmp/test_nccl_trace_rank_", rank);
```
doesnt the filename also need be read from the `TORCH_NCCL_DEBUG_INFO_TEMP_FILE` env variable?

XilunWu replied:
I think for testing we can use what Howard suggests -- `tmpfile`.

fduwjj replied:
@H-Huang yes, for test purpose, I just skipped it. I will added it in the test. And I will use `tmpfile` here @XilunWu

XilunWu commented on test/cpp/c10d/ProcessGroupNCCLErrorsTest.cpp:
```diff
@@ -346,23 +347,64 @@ TEST_F(ProcessGroupNCCLErrorsTest, testNCCLErrorsNonBlocking) {
+        // std::strncpy or std::strcpy does not work here because there is
+        // null terminator character (\0) in the string.
+        traces.assign(ncclTrace.begin(), ncclTrace.end());
+        file.write(reinterpret_cast<const char*>(&size), sizeof(size));
+        file.write(ncclTrace.data(), size);
```
shall we call `file.flush()` after writing?

fduwjj replied:
actually, since right after this, the destructor of `ofstream` will be called. The `flush` will be called automatically.

fduwjj committed:
Update on "[NCCL PG] Enable storing nccl traces into storage and make it configurable"


This PR is to enable the store of NCCL flight recorder to storage and make it configurable by letting users register their own way of storing the debug info. We will then provide users a script to offline parse and process the dumped blobs.

One thing, this PR is not trying to resolve is to decide where to dump the debug info. I will send a follow-up PR to address that.


cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc

[ghstack-poisoned]

fduwjj committed:
Update on "[NCCL PG] Enable storing nccl traces into storage and make it configurable"


This PR is to enable the store of NCCL flight recorder to storage and make it configurable by letting users register their own way of storing the debug info. We will then provide users a script to offline parse and process the dumped blobs.

One thing, this PR is not trying to resolve is to decide where to dump the debug info. I will send a follow-up PR to address that.


cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc

[ghstack-poisoned]

fduwjj committed:
Update on "[NCCL PG] Enable storing nccl traces into storage and make it configurable"


This PR is to enable the store of NCCL flight recorder to storage and make it configurable by letting users register their own way of storing the debug info. We will then provide users a script to offline parse and process the dumped blobs.

One thing, this PR is not trying to resolve is to decide where to dump the debug info. I will send a follow-up PR to address that.


cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -313,6 +313,16 @@ class TORCH_API ProcessGroupNCCL : public Backend {
     friend class ProcessGroupNCCL;
   };
 
+  class DebugInfoCallbackStorer {
```
nit: maybe a shorter name?
What about DebugInfoWriter

fduwjj replied:
Sure.. Let's use the name Writer, originally I am not quite sure if we need to put FallBack in the name.

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -313,6 +313,16 @@ class TORCH_API ProcessGroupNCCL : public Backend {
+   public:
+    DebugInfoCallbackStorer(std::string fileName);
+    virtual ~DebugInfoCallbackStorer();
+    virtual void storeTraceInfoStorer(int rank, const std::string& ncclTrace);
```
3 questions about this API

1) should rank be an arg? (why?) (and if so why not put it into ctor)
2) should ncclTrace be an std::string or do we want to make it another type?  (bytes, or, actual structure?)
3) nit: rename?  just simply "write()"?

fduwjj replied:
We want the file comes with rank in its name? I mean for examples, if 8 ranks on a host to write to one single file, we will have race condition and the doc will be in a uncertain state (we don't know which rank will write first, right?)

ncclTrace is actually a vector of <uint_8>, I think @zdevito use string just let users to treat them as a string blob, so that users will do `pickle.loads` in python to convert it back to a list of map I believe. Sure, let's use the name of `write()`

zdevito replied:
We should change that api to actually return std::vector<uint8_t> so that it is less confusing. I know it really only is being used in our code, but it's come up a number of times. I just did `std::string` because it was quick.

fduwjj replied:
Sure, I will send a separate PR to change it.

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1276,13 +1302,33 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
 }
 
-void dump_debugging_info() {
+void ProcessGroupNCCL::registerDebugInfoCallbackStorer(
```
CallbackStorer -> Writer?

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1276,13 +1302,33 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
+    auto ncclTrace = dump_nccl_trace();
+    if (debugInfoCallbackStorer_ == nullptr) {
+      // Dump the trace blob into local disk as a fallback.
+      const char* fileName = parseEnvVarString(
```
its a little odd to me to "just in time" parse the env and construct the default writer.  i wonder why not create it inside the constructor?  Do we benefit from having ability to change the writer mid-way through the program?

fduwjj replied:
Actually I don't have a strong opinion on this. To be passing in the constructor vs getting the path name in the constructor is pretty much the same.

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1341,13 +1385,21 @@ void ProcessGroupNCCL::heartbeatMonitor() {
   // kill the whole process.
-  if ((terminateProcessGroup_.load() || collectiveDebugInfoMode_.load()) &&
+  if ((terminateProcessGroup_.load() || collectiveDebugInfoMode_.load() ||
+       !debugInfoStoreThread.joinable()) &&
```
is it possible joinable is false here? under which conditions?

fduwjj replied:
When the thread has not finished or not initialized, `joinable` return False.

fduwjj replied:
Hmm this condition should be `debugInfoStoreThread.joinable()` I got confused `joinable` with the end of thread execution previously..

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1341,13 +1385,21 @@ void ProcessGroupNCCL::heartbeatMonitor() {
+  // At this point, we either already sleep for another `heartbeatTimeoutInSec_`
+  // or the thread has finished. Because we don't want to block the monitor
+  // thread, so We mark the thread detach and the dump of debug info becomes
+  // "best effort". If the process exit normally, marking it detach also makes
```
if the process exits normally, is it possible for the heartbeat monitor to reach this line? if we ever exit the loop above, does it mean an error happened?

fduwjj replied:
If the process exits normally, we will early return, then no need to dump the debug info right? And yes, it's still possible, but you are right, I need to add a `joinable()` check, I was under the wrong impression that we can detach a default constructed thread..

wconstab commented on test/cpp/c10d/ProcessGroupNCCLErrorsTest.cpp:
```diff
@@ -346,6 +346,55 @@ TEST_F(ProcessGroupNCCLErrorsTest, testNCCLErrorsNonBlocking) {
+      // One thing which can not be explained is that, without this line of log,
+      // the last element of `tracesFromStorer_` will be different from
+      // `tracesToStorer_`. Since using TmpFile is just for testing purpose,
+      // this is probably not a big deal.
```
better explain it, is it due to our thread exiting before it flushes its file? do we need to flush again before we return?

fduwjj replied:
Actually I don't think it is due to thread exiting because we do the write right after read. And we are reading here, we don't need to call flush, right? The weird thing is that if we comment the line of LOG(INFO), the comparison of two vector will be different, which I cannot reason how come it is like that.

fduwjj replied:
When using stream and filesystem, this issue is gone..

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -907,6 +908,31 @@ bool ProcessGroupNCCL::CoalescedWorkNCCL::wait(
+    return;
+  }
+
+  file.write(ncclTrace.data(), ncclTrace.size());
```
wonder if we're gonna need the same solution for flushing this file, as we need for the test case.

in fact, i wonder if you even need to subclass the writer.  can you just open the tempfile inside the test case and then pass the path to the tempfile to the constructor for this writer, and fully reuse it?

fduwjj replied:
Actually since we are using streaming and it will automatically flush one the function exit. Unfortunately, `std::tmpfile` does not support a given path... Read/write using stream is much better than using fgets/fputs, but looks like std::tmpfile does not work with stream either.. One alternative here is that we manually manage the file removal ourselves so that we can generate a temporary file name using `std::tmpnam` and use std::remove in the end to remove the file. WDYT?

fduwjj replied:
Actually, from cpp 17, there is something called std::filesystem; we can use it to replace `std::tmpfile`. But in the end, we still need to do something like: `fs::remove(tempPath);`

fduwjj replied:
OK, I am able to just add a slight change to the write and reuse most of logic from the default writer. Also we can still use the path set from Env variable by using `std::filesystem::temp_directory_path` and `std::filesystem::remove`.

fduwjj committed:
Update on "[NCCL PG] Enable storing nccl traces into storage and make it configurable"


This PR is to enable the store of NCCL flight recorder to storage and make it configurable by letting users register their own way of storing the debug info. We will then provide users a script to offline parse and process the dumped blobs.

One thing, this PR is not trying to resolve is to decide where to dump the debug info. I will send a follow-up PR to address that.


cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc

[ghstack-poisoned]

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -313,6 +313,16 @@ class TORCH_API ProcessGroupNCCL : public Backend {
     friend class ProcessGroupNCCL;
   };
 
+  class DebugInfoWriter {
```
Move this class to `NCCLUtils.hpp` or even `Utils.hpp`? (depends on how general it is)

fduwjj replied:
I am not sure how much we benefit from moving it to `Utils.hpp` maybe `NCCLUtils.hpp`?

zdevito replied:
Is there a reason this can't just be a `std::function<void(std::string&)>`?

fduwjj replied:
Actually, @wconstab and I had a conversation above. Originally I was using the function in the code but making a class can make it more generic and easier for internal to override?

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -907,6 +908,30 @@ bool ProcessGroupNCCL::CoalescedWorkNCCL::wait(
+    return;
+  }
+
+  file.write(ncclTrace.data(), ncclTrace.size());
```
Is this write behavior on by default or gated by an env? We should not directly write logs into user's disk by default.

fduwjj replied:
It's gated and default behavior is for OSS users, advanced users will override the write behavior.

kwen2501 replied:
Sorry which env gates it? I might have missed it.

fduwjj replied:
If users set `TORCH_NCCL_TRACE_BUFFER_SIZE` larger than 0, users will enable the flight recorder. `TORCH_NCCL_ENABLE_MONITORING` will enable the dumping and monitor thread.

kwen2501 replied:
Just a thought, not necessary for this PR: we can rely on whether `TORCH_NCCL_DEBUG_FILE` is set to enable/disable writing.

fduwjj replied:
yeah, we can do that. But if you look at the implementation of the default write, we don't write the the file if `!file.is_open()`.

zdevito reviewed (approved):
Looks good! I have one minor nit, but I think we should land this to unblock deadlock investigations, we can sort out nits later.

fduwjj committed:
Update on "[NCCL PG] Enable storing nccl traces into storage and make it configurable"


This PR is to enable the store of NCCL flight recorder to storage and make it configurable by letting users register their own way of storing the debug info. We will then provide users a script to offline parse and process the dumped blobs.

One thing, this PR is not trying to resolve is to decide where to dump the debug info. I will send a follow-up PR to address that.


cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge

fduwjj commented:
@pytorchbot merge -i
